src/bench/crypto_hash.cpp

```diff
@@ -190,9 +191,19 @@ static void SHA512(benchmark::Bench& bench)
 static void SipHash_32b(benchmark::Bench& bench)
 {
     uint256 x;
-    uint64_t k1 = 0;
+    uint64_t k0 = 0, k1 = 0;
     bench.run([&] {
-        *((uint64_t*)x.begin()) = SipHashUint256(0, ++k1, x);
```

Comment:
@fanquake, is it possible that [this way of benchmarking](https://github.com/bitcoin/bitcoin/commit/619d5691c20bee5d08be2ce85aafa2cb570dbca4#diff-722a565cc61ff748b555e7ed62c0affd13618f67c693524edc351c7b4bf20247R67) isn't actually representative?
I tried comparing code that was obviously faster according to the benchmark by running the before and after code explicitly many times in a test and measuring total time - and the new version seems to reflect it more accurately.
I tried it with `ankerl::nanobench::doNotOptimizeAway` as well, but the results seem to be exactly the same.

Reply:
@sipa, can you please confirm that the previous benchmark was indeed incorrect?

Reply:
Yes, I understand the reason, that's why `doNotOptimizeAway` is needed in some cases.
But here it seems to produce the wrong result. I have validated it with a simple timed test, calculating a million different hashes with the two versions of the algorithm, which were giving different benchmarking speeds - and this new version resembles it more closely.
Also, we're not using this format in any other benchmarks I've seen.

Reply:
You're right, that's why it's still a draft. Thanks for the early input.

Reply:
@maflcko, so my point was that it seems to me that changing the benchmark slightly changes it's performance considerably:
```C++
static void SipHash_32b(benchmark::Bench& bench)
{
    uint256 x;
    uint64_t k1 = 0;
    bench.run([&] {
        *((uint64_t*)x.begin()) = SipHashUint256(0, ++k1, x);
    });
}
```

works with inputs such as:
<img src="https://github.com/bitcoin/bitcoin/assets/1841944/66d3022e-fa15-4950-bf12-c7db9896d766">
and the benchmark results in:
> make -j10 && ./src/bench/bench_bitcoin --filter='SipHash_32b' --min-time=10000

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               35.11 |       28,479,847.20 |    0.1% |     11.00 | `SipHash_32b`
-------

Changing the benchmark by adding starting values for each input and modifying every 64 bit chunk of x, and consuming the `SipHashUint256` result via `doNotOptimizeAway`, as follows:
```C++
static void SipHash_32b_new(benchmark::Bench& bench)
{
    FastRandomContext rng(true);
    auto k0{rng.rand64()}, k1{rng.rand64()};
    auto x{rng.rand256()};
    auto* x_ptr{reinterpret_cast<uint64_t*>(x.data())};
    bench.run([&] {
        ankerl::nanobench::doNotOptimizeAway(SipHashUint256(k0, k1, x));
        ++k0; ++k1; ++x_ptr[0]; ++x_ptr[1]; ++x_ptr[2]; ++x_ptr[3];
    });
}
```

which would work with inputs such as:
<img src="https://github.com/bitcoin/bitcoin/assets/1841944/5470f279-d3a9-444c-8460-b5febe3f1379">
results in the following benchmark:
> make -j10 && ./src/bench/bench_bitcoin --filter='SipHash_32b_new' --min-time=10000

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               21.54 |       46,420,932.76 |    0.1% |     10.98 | `SipHash_32b_new`

-----
I've added `doNotOptimizeAway` to every other benchmark in `crypto_hash.cpp` and their values didn't change considerably.

For the record, the following is also running a lot faster:
```C++
static void SipHash_32b_new(benchmark::Bench& bench)
{
    uint256 x;
    uint64_t k1 = 0;
    bench.run([&] {
        auto result = SipHashUint256(0, ++k1, x);
        ankerl::nanobench::doNotOptimizeAway(result);
        *((uint64_t*)x.begin()) += 1;
    });
}
```
but adding `result` instead of the `1` makes it slow again:
```C++
        *((uint64_t*)x.begin()) += result;
```

Reply:
Can you please be more specific that this?
My intention is to measure `SipHash` more realistically, since it didn't behave as I expected before.

Reply:
Thanks @sipa, do you agree that
```
static void SipHash_32b(benchmark::Bench& bench)
{
    FastRandomContext rng(true);
    auto k0{rng.rand64()}, k1{rng.rand64()};
    auto x{rng.rand256()};
    auto* x_ptr{reinterpret_cast<uint64_t*>(x.data())};
    bench.run([&] {
        ankerl::nanobench::doNotOptimizeAway(SipHashUint256(k0, k1, x));
        ++k0; ++k1; ++x_ptr[0]; ++x_ptr[1]; ++x_ptr[2]; ++x_ptr[3];
    });
}
```
likely produces more realistic results (see above results)?
The current one gave me a few false optimization hopes - which this one didn't?

Reply:
Thank you, I've pushed it in https://github.com/bitcoin/bitcoin/pull/30349